### ISSUE TITLE
Required PHP version

### ISSUE THREAD

**Issue by sebastianbergmann** [sebastianbergmann/phpunit#1595]
PHPUnit is currently supported on PHP 5.3.3 and newer. PHP 5.3.3 was [released](http://php.net/ChangeLog-5.php#5.3.3) in July 2010 and [stopped being supported](http://php.net/eol.php) by the PHP project in August 2014. PHP 5.6 was [released](http://php.net/ChangeLog-5.php#5.6.0) in August 2014 and [is supported](http://php.net/supported-versions.php) by the PHP project until August 2016.

The [release process](https://github.com/sebastianbergmann/phpunit/wiki/Release-Process#php-version-requirement) of PHPUnit requires that a PHPUnit release series that is actively supported by the PHPUnit project must be compatible with all versions of PHP that are actively supported by the PHP project. It also states that the minimum required version of PHP may only be bumped in a new major release of PHPUnit.

I would like to bump the required version of PHP to PHP 5.6 for PHPUnit 5 in an effort to lessen the burden to support outdated PHP versions for the PHPUnit development team and to motivate the PHP community to adopt newer versions of PHP. PHPUnit 5 should be the last major version of PHPUnit to support PHP 5. PHPUnit 6 should only support PHP 7.

[I do not want to leave users of PHPUnit behind.](http://blog.ircmaxell.com/2014/12/on-php-version-requirements.html) However, there are quite a few language features in newer versions of PHP, especially in PHP 7, that I would like to leverage for the development of PHPUnit. According to PHPUnit's [release process](https://github.com/sebastianbergmann/phpunit/wiki/Release-Process#support), PHPUnit 4 would be supported through bugfix releases for at least one year after PHPUnit 5 comes out.

With regards to a timeline, I am currently thinking of releasing PHPUnit 5 in December 2015 and PHPUnit 6 in December 2016.

**Comment by TomasVotruba** [comment]
I agree, it's according to release process, so I don't see any problem with that.
Also, actively supported PHP versions [will be PHP 5.6+](http://php.net/supported-versions.php) in December 2015.

Also this is great inspiration for other packages. Thanks for stepping up!

**Comment by nilopc** [comment]
PHPUnit 6 is a must. 

PHPUnit 5 sounds good with 5.6, but maybe you should target the lowest official supported version if you're concerned about not leaving users behind. So I'd go from 5.4 to 5.6 gradually, whenever they ditch 5.4 move on to 5.5, etc.

**Comment by flijten** [comment]
Basically there's two ways to look at this. From the perspecive of adoption and getting as much people to run a supported version of PHP I'd shout yes wholeheartedly. From the perspective of getting as much people to automatically test their code moving upwards with your required PHP version is a problem .

So what would I do? I'd definitively go with new required versions. I'd wager that most people who care about their code enough to have it tested would also be caring enough to go with newer PHP versions. And if they're hosting somewhere where newer PHP versions are not an option they'd care enough to go somewhere else.

**Comment by martinssipenko** [comment]
:+1:

**Comment by Ocramius** [comment]
The current feature set is enough for existing PHP versions, IMO.

**Comment by hhamon** [comment]
Totally agree!

**Comment by jk** [comment]
> So what would I do? I'd definitively go with new required versions. I'd wager that most people who care about their code enough to have it tested would also be caring enough to go with newer PHP versions. And if they're hosting somewhere where newer PHP versions are not an option they'd care enough to go somewhere else.

Basically I agree, but you should not forget about those people how do enterprise stuff which depends on vendor-specific PHP versions (i.e. CentOS, Ubuntu LTS, etc.) where they just moved away from PHP 5.3.3 to PHP 5.4 - and that's here to stay for while (rather years than months).

**But**: IMO they can stay on PHPUnit 4.x as a composer dependency. Composer definitively took the burden of dependency management even for development deps. So: Go for the planed migration path.

**Comment by agiuliano** [comment]
totally agree!

**Comment by Keloran** [comment]
sounds like a good idea

**Comment by giorrrgio** [comment]
:+1:

**Comment by smoya** [comment]
We know that keep an older version of your stack can carry this kind of situations. Anyway, IMHO giving support to PHP5.3 is like a website giving it to IE6 (or 7?)
So, +1

**Comment by garak** [comment]
:+1:

**Comment by eux** [comment]
:+1:

**Comment by hanneskaeufler** [comment]
Go for it. If you´re not on the shiny new php, you´re not getting the shiny new phpunit. :+1:

**Comment by MAXakaWIZARD** [comment]
+1

**Comment by localheinz** [comment]
:+1:

**Comment by sarelvdwalt** [comment]
I agree, like hanneskaeufler said, go for it. You already stated that bugfixes will be done, and that's fine, but new functionality and awesome tools on the newest platform is something most people are comfortable dealing with.

**Comment by henriquemoody** [comment]
I totally agree.

Since PHPUnit 4 will be supported for a year I see no problem on that. Of course uses who need to use these older PHP versions will not get new features in PHPUnit but they already don't new features in the PHP in itself.

There are some pieces of code on PHPUnit (and related libraries) with workarounds to keep compatibility between versions and with this change I think it would be better for developers make improvements.

**Comment by sun** [comment]
Not sure I understand the proposed plan — perhaps I'm missing something?

> bump the required version of PHP to PHP 5.6 for PHPUnit 5
> PHPUnit 5 should be the last major version of PHPUnit to support PHP 5.
> PHPUnit 6 should only support PHP 7.

To my knowledge, all future PHP 5.* series releases will only contain bugfixes but no major changes (5.7 is scheduled to be the last minor release; it only adds new deprecation warnings).

Thus - in terms of support - there is nothing crazy to expect from PHP 5.* at this point anymore.

Likewise, I'm not aware of a single case in PHPUnit support that would have ended with the conclusion, _"Oh, you need PHP 5.6 or later for that."_

So, Why would we want to create an intermediate major version, _only_ to get rid of PHP 5.3 - 5.5?

Just to visualize the proposed plan:

| PHPUnit | PHP.min() | PHP.max() | What's new?
| --- | --- | --- | --- |
| 4 | 5.3 | 5.* | –
| 5 | 5.6 | 5.* _(or 7?)_ | Requires PHP 5.6
| 6 | 7.* | 7.* | Great, fancy new stuff, requires PHP 7

Unless I missed something big, there is no major change in functionality in the current pipeline that couldn't be done in the PHPUnit 4 series.  There are a few refactoring ideas (and branches), but they are far from being ready; also, their changes are much rather material for a new major version that is _really major_ (similar to requiring and leveraging PHP 7).

I'm left with the impression that PHPUnit 5 wouldn't ship with anything new (as you'd expect from a new major version) and would only exist to drop support for "older" PHP versions (whereas it will require a newer PHP version that may not be available on OS LTS releases and hosting infrastructure).

Why don't we skip the intermediate major version in between, and instead, focus on making PHPUnit 5 a true major release, requiring PHP 7, with plenty of major improvements, possibly even requiring users to adapt their tests?

**Comment by mgonyan** [comment]
:+1:

**Comment by sebastianbergmann** [comment]
The problem with major releases is that they are "major". I want to keep the flow of releases that add new features every two months. What would be wrong with the only break between PHPUnit 4 and PHPUnit 5 would be the PHP version requirement?

I do plan, however, to have PHPUnit itself use namespaces in time for PHPUnit 5. So we would have that in PHPUnit 5 (with a BC layer that makes extending from `PHPUnit_Framework_TestCase` still work, for instance, in place). And then a year later with PHPUnit remove that BC layer.

**Comment by andreasemer** [comment]
:+1:

**Comment by GrahamCampbell** [comment]
> I do plan, however, to have PHPUnit itself use namespaces in time for PHPUnit 5.

Namespaces, PSR-4, and use of the actual new stuff php has would be cool. Much of phpunit's code is still written in such a way it would work on php 5.2 (if you get what I mean - not referring to the old style "namespacing").

**Comment by Intrepidity** [comment]
:+1:

**Comment by henriquemoody** [comment]
I'm with @GrahamCampbell on that.

Since namespaces are going to be used and it's a huge API break, I think it is a good time to make improvements, like #913 and #1355 and [other refactoring issues](https://github.com/sebastianbergmann/phpunit/issues?q=is%3Aopen+is%3Aissue+label%3Arefactoring).

**Comment by GrahamCampbell** [comment]
@henriquemoody I think it would be worth just doing a phpunit 5 that requires php 7 and make it totally awesome. It gives you much more time.

**Comment by lcobucci** [comment]
I totally agree with that, and we will have a lot of good benefits.

**Comment by rdohms** [comment]
:+1:

**Comment by SenseException** [comment]
Are versions < 6 and >= 5 supporting PHP 7?

**Comment by sebastianbergmann** [comment]
PHPUnit 4 "just works" on PHP 7.

**Comment by sebastianbergmann** [comment]
The reason I propose to bump to PHP 5.6 first is because of the [release process](https://github.com/sebastianbergmann/phpunit/wiki/Release-Process#php-version-requirement) of PHPUnit requires that a PHPUnit release series that is actively supported by the PHPUnit project must be compatible with all versions of PHP that are actively supported by the PHP project. The active support for PHP 5.5 ends in June 2015. By the time PHPUnit 5 comes out PHP 5.6 will be the only version of PHP actively supported.

We could make the timeline more aggressive, though, and drop support for PHP versions < PHP 5.6 and release PHPUnit 5 in June 2015.

**Comment by sebastianbergmann** [comment]
Thinking on this further, we could introduce an automatism: when PHP drops active support for a version of PHP then PHPUnit drops support for that version of PHP. When a new version of PHP is released then PHP needs to add support for that version.

I like it. Thoughts?

**Comment by sebastianbergmann** [comment]
@sun @henriquemoody A version number is just that: a number. [Semantic Versioning](http://semver.org/spec/v2.0.0.html) states that the major version has to be incremented "when you make incompatible API changes". As much as I agree with Semantic Versioning I think it has a shortcoming in not requiring to increment the major version when you drop support for a version of the platform (in our case PHP) your software (in our case PHPUnit) is built on. That is why the release process of PHPUnit states that the minimum required version of PHP may only be bumped in a new major release of PHPUnit.

**Comment by henriquemoody** [comment]
I think PHPUnit must just drop the support for a version when PHP drops the _security fixes only_ support and not when the _active support_ is dropped for the version.

IMO, doesn't make much sense to keep support for a version of PHP which is not supported for PHP in itself, unless to keep this support is not a problem. PHP 5.3 is not longer supported by PHP and there are lots of libraries that still support it because don't need of features of 5.4>. If is that the case, PHPUnit can keep the support for later versions.

Any way, the PHPUnit's release process may not match with PHP's release process.

**Comment by sebastianbergmann** [comment]
PHPUnit releases a new version, either `X.Y+1.0` or `X+1.0.0`, every two months. PHP releases a new version (`X.Y+1.0`) every year. Aligning the processes should not pose a problem.

**Comment by henriquemoody** [comment]
When I was thinking about the _API break_ I was not necessary thinking about Semantic Versioning, but switch to namespaces is a huge BC break, doesn't matter what version number it is.

You said the `PHPUnit_Framework_TestCase` class will be kept but how about the internal classes? As far I know, the issues I mentioned are not easy to be concluded without break BC.

Drop the support for older versions and switch to namespaces is a awesome thing, even if you do nothing more than that, but since it's happening I think it's good moment to think about them.

**Comment by marcioAlmada** [comment]
TLDR; In favor of moving forward, specially if move forward means to target PHP7+

For the people concerned with BC breaks. Don't forget PHP7 has return type hints [now](https://github.com/php/php-src/commit/c8576c5a46508ce87f7d2eb2442e6a9283b6a1ce) so a phpunit that targets PHP7+ and make use of the new added features makes sense. Besides that, users that decide to stick with PHP5 will still have phpunit LTS versions available for a long time.

:+1: :+1: :+1:

**Comment by eddiejaoude** [comment]
> ...I would like for PHPUnit 5 to require PHP 5.6 and PHPUnit 6 to require PHP 7...

sounds good :+1:

**Comment by sun** [comment]
As a software testing framework, PHPUnit should not attempt to strictly follow the fast-paced release plan of the upstream PHP language, but instead, it should be oriented at the slow-paced adoption and system requirements of the downstream projects/consumers that are using PHPUnit to test their software products.

A PHPUnit 5 that requires PHP 5.6 cannot be used by many/most consumers, because _their_ requirements will still support PHP 5.4 or similar.  Remember, it's just a _minimum_ version compatibility, not a requirement for people to actually run on that version.  Downstream projects won't be able to use and adopt PHPUnit 5 if they cannot run their tests on all of _their_ supported versions.

For that reason, a testing framework should try to retain compatibility as long as possible (and sensible).  The only case I'd understand is the PHP 7 case, ideally combined with some major refactoring and rewrite work, so as to address some truly major architectural design topics in a new major version of PHPUnit.

> What would be wrong with the only break between PHPUnit 4 and PHPUnit 5 would be the PHP version requirement?

From my perspective, it's the exact case of dropping compatibility/support without having an important reason, as discussed in [@ircmaxell's post](http://blog.ircmaxell.com/2014/12/on-php-version-requirements.html).  It becomes clear by reversing the question:

What would be wrong with supporting PHP 5.4 for a little longer?

> use namespaces for PHPUnit 5 (with a BC layer)

As far as I can see, it is possible to do that in a new minor version of PHPUnit 4 already?

> Semantic Versioning… has a shortcoming in not requiring to increment the major version when you drop support for a version of the platform your software is built on.

While I agree with you on theoretical grounds, it's not feasible in practice.  Primarily for economical and ecosystem reasons, the software release life cycle of downstream software projects (end products) is usually much longer than the duration/availability of a PHP minor version (obviously even more so for LTS releases).  Therefore, many projects are removing compatibility and support for obviously obsolete or insecure versions in new minor versions.  PHPUnit could follow that approach.

Would anyone complain if support for end-of-life PHP 5.3 (only) would be removed in the third to next major version PHPUnit 4.7.0 (around September 2015)?

I don't think so.  Users who must still test against PHP 5.3 for whichever weird reasons can stay on 4.6.x until they have resolved their dependencies.

In short, I'd recommend to focus on major changes and improvements for PHPUnit 5.  Only if deemed really necessary, drop support for PHP 5.3 in PHPUnit 4.7 or later.

**Comment by sebastianbergmann** [comment]
I do not want to bump the required PHP version in a minor version such as PHPUnit 4.7. And I have no problem with incrementing PHPUnit's major version just for bumping the required PHP version.

**Comment by yguedidi** [comment]
> I would like to bump the required version of PHP to PHP 5.6 for PHPUnit 5 ... PHPUnit 6 should only support PHP 7.

Of course: Yes!

**Comment by HallofFamer** [comment]
Umm I thought PHP Unit still supports PHP 5.2 as I still see class name like PHPUnit_Framework_TestCase rather than PHPUNIT\Framework\TestCase. I thought PHP Unit refuses to use namespace since it supports PHP 5.2? Or am I missing something here?

**Comment by whatthejeff** [comment]
@HallofFamer See https://phpunit.de/manual/current/en/installation.html#installation.requirements.

**Comment by HallofFamer** [comment]
@whatthejeff: Thanks for your comment, but if PHPUnit already drops PHP 5.2 support, why it is not using namespace yet?

**Comment by Ocramius** [comment]
Mainly because the change would not be required. It's an un-needed BC Break.

**Comment by whatthejeff** [comment]
@HallofFamer: @Ocramius is correct. Some of PHPUnit already [uses namespaces](https://github.com/sebastianbergmann/phpunit/blob/2e8580deebb7d1ac92ac878595e6bffe01069c2a/src/Framework/TestCase.php#L11-L15), but BC is the main reason we haven't transitioned the entire library yet.

**Comment by JeroenDeDauw** [comment]
+1

**Comment by HallofFamer** [comment]
Okay I understand, so will PHPUnit be using Namespace extensively in version 5? I mean, will things like PHPUnit_Framework_TestCase become PHPUnit\Framework\TestCase in PHPUnit 5.0? Or is this planned for 6.0 instead? or will it not happen at all?

**Comment by whatthejeff** [comment]
> Okay I understand, so will PHPUnit be using Namespace extensively in version 5? I mean, will things like `PHPUnit_Framework_TestCase` become `PHPUnit\Framework\TestCase` in PHPUnit 5.0?

@HallofFamer Yep, see https://github.com/sebastianbergmann/phpunit/issues/1595#issuecomment-73578453.
